### DISCUSSION TITLE
Expected behavior of comprehensions

### DISCUSSION THREAD

**Discussion by mingweishih** [open-policy-agent/community#323] [OPA and Rego] [answered]
I was trying to understand how the comprehensions work. I tested the following example in the Rego Playground:
```
array1 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

array2 = [[1, 2, 3], [4, 5, 4], [7, 8, 9]]

compr = [ 1 | array1[_] == array2[_] ]
```

The results of `compr` is 
```
    "compr": [
        1,
        1
    ]
```

The likely explanation is that the `<body>` does the following
```
array1[0] == array2[0] // true => generate 1
array1[1] == array2[1] // false => generate nothing
array1[2] == array2[2] // true => generate 1
```

But I can't find references that explain whether this is correct. Is this an expected behavior?

**Comment by anderseknert** [comment, marked as answer]
Hi there!

Your analysis is correct. There are probably better explanations available in the docs, but [this one](https://www.openpolicyagent.org/docs/latest/policy-language/#the-basics) comes to mind.

> When evaluating rule bodies, OPA searches for variable bindings that make all of the expressions true. There may be multiple sets of bindings that make the rule body true.

The same goes for comprehensions — there are two "bindings" in your comprehension where the expression is true, resulting in two items stored in the resulting array.

**Reply by mingweishih** [reply]
Thanks for the clarification!
